overseer/src/lib.rs

```diff
@@ -537,6 +564,13 @@ where
 
 	/// Run the `Overseer`.
 	pub async fn run(mut self) -> SubsystemResult<()> {
+		let leaves = std::mem::replace(&mut self.leaves, vec![]);
+
+		for leaf in leaves.into_iter() {
```

Comment:
`leaves.drain(..)` would make the above line redundant

Reply:
look like that would need two mutable refs to `self`?

Reply:
Does this work?

```rust
for leaf in self.leaves.drain(..) {

}

// borrow of `self` should not last here, but if rustc thinks it does, we can wrap in an enclosing scope
```

Reply:
```rust
		for leaf in self.leaves.drain(..) {
			self.broadcast_signal(OverseerSignal::StartWork(leaf)).await?;
			self.active_leaves.insert(leaf);
		}

```

these are two mutable refs: the `drain` is the first mutable refs and calls to `self` in the body of the loop are other mutable refs.

Reply:
I see. Then Basti's suggestion of `mem::take` makes most sense. That way we will deallocate the underlying storage, unlike in my suggestion.